audbcards/core/dataset.py

```diff
@@ -28,10 +56,12 @@ class Dataset:
         cache_root: cache folder
 
     """
+    @datasetcache
```

Comment:
> I guess the `@datasetcache` decorator is needed as we want to load the whole object inside `__init__.py` instead of executing the lines in `__init__.py`? Will it automatically skip all those steps?

Yes this was the intention. Imo the old implementation using the decorator was not ideal though: the decorator approach seems to be often used in scenarions dealing with [caching single properties](<https://stackoverflow.com/questions/815110/is-there-a-decorator-to-simply-cache-function-return-values>), but rather not in situations dealing with object construction. As this posts suggests, a better solution is to use a [constructor implemented as a class method](<https://stackoverflow.com/questions/13054250/python-object-cache>). This even would to instantiate objects in different fashions (constructor polymorphism), but for us, it separates out the object caching into a separate method. It comes at the cost of calling the constructor, i.e. `ds = Dataset(*args, **kwargs)` becomes `ds = =Dataset.create(*args, **kwargs)`. In summary: putting everything into `__init__` is also possible, however the [link above](<https://stackoverflow.com/questions/13054250/python-object-cache>) is recommending against it. Do you have a strong oppinion?

Reply:
I would be in favor of still using `Dataset()`. So maybe the solution including `__new__` works as proposed in https://stackoverflow.com/a/13054570 (even though I don't understand it yet).

Reply:
Before you spend too much effort to avoid the `__init__()` step it might be worse to measure how much time we can actually safe. The main stuff we are doing here is loading the database dependencies (which is cache by `audb` as a pickle file) and loading the database header file (which is always loaded from a YAML file in `audb`).
I think we can get faster by caching, but maybe we should be sure and do some measurements.

Reply:
I am have been having problems to get this appproach to run without a class method type constructor:

This should work:

```python
class Dataset:
    r"""Dataset.

    Dataset object that represents a dataset
    that can be loaded with :func:`audb.load()`.

    Args:
        name: name of dataset
        version: version of dataset
        cache_root: cache folder

    """
    def __new__(cls,
                name: str,
                version: str,
                *,
                cache_root: str = '~/.cache/audbcards',
    ):

        dataset_cache_filename = cls._dataset_cache_path(name,
                                                         version,
                                                         cache_root)

        if os.path.exists(dataset_cache_filename):
            print(f"loading from cache: {dataset_cache_filename}")
            obj = cls._load_pickled(dataset_cache_filename)
            obj._from_cache = True
            return obj

        # return the newly constructed object
        return cls.__init__(cls, name, version, cache_root)
```

The only other modification needed is to make `__init__` to return `self`

However this will return the class and not an instance.  I will try to modify the constructor to use the classmethod constructor, but leaving the signature intact (i.e. without the call to the `create`  method).

Reply:
> Before you spend too much effort to avoid the `__init__()` step it might be worse to measure how much time we can actually safe. The main stuff we are doing here is loading the database dependencies (which is cache by `audb` as a pickle file) and loading the database header file (which is always loaded from a YAML file in `audb`). I think we can get faster by caching, but maybe we should be sure and do some measurements.

I believe it is always there now. But there is an important point to this anyway: calling `Dataset.properties` that effectively collects all properties is taking fairly long, for example, `audiosets` is really slow to get hold of.

Reply:
OK, then maybe let's indeed first create some benchmarks and see how much faster we are using `Dataset.create()` and the cache, opposed to the implementation in `main` which simply relies on `audb` for the caching when calling `Dataset().__init__()`.

Reply:
I have measured the creation of uncached and cached instance creation  using the `@classmethod `  constructor as implemented in this MR for an array of 170 databases.  On average this is a speed-up of a factor between 5 to 6 when comparing cached versus uncached instance creation.

![image](https://github.com/audeering/audbcards/assets/14260675/35bfbea7-69c2-4640-98f6-48fabbd8e3bb)

In addition, I have measured the time to extract all properties from the data. This seems to be the more severe problem with again being a factor 30 slower than the uncached creation. 

It would very likely make sense to identify the "slow" properties as most performance gain potential sits there, and think about how to deal with it. If you create your dataset properties only once then caching properties to memory is not great of a help as you suggested earlier, and one has to think about alternatives. 

Btw. what were your reservations against a `@classmethod` type constructor again? It seems to be the preferred approach on the internet, even used by hightly acclaimed packages like pandas, e.g [`pandas.DataFrame.from_dict`](
https://github.com/pandas-dev/pandas/blob/fe178189e589711559a3064fcefa4d330b7f41fa/pandas/core/frame.py#L1741) . While it surely is more interesting to think about the time to spend  to calculate the properties and how to cache these, 5 versus 32 seconds as the case above is already a gain.

Reply:
My reservations against `@classmethod` is that the user should not have to take care if a version is cached or not, but should just use always the same command. If this is not possible with calling `Dataset()`, we can maybe replace it by a function call, e.g. `dataset()` that then internally takes care of calling `Dataset()` or `Dataset.create()`. This might still look a little bit strange in the documentation and a user might wonder why not directly call `Dataset()`, but at least it's always the same command.

Reply:
Great to see that caching brings significant boosts in both cases. So it seems a good idea to cache the object, and with maybe greater priority single properties.

For caching properties there exists https://docs.python.org/dev/library/functools.html#functools.cached_property, but this caches only per instance. As in our case the properties will never change for an existing database version, it will make more sense to directly cache them on disc. Maybe we could write a custom wrapper for it in the style of https://stackoverflow.com/q/59929626 and add code there to pickle and read from pickle.

Reply:
For caching: 

I have played with it and think the `cached_property` is helpful. I have implemented it in `ce73026`. You do not even have to calculate the property if you want to push the properties into the cache - using the constructor interestingly seems to be enough.

I have added a test as I didn't believe it. The memory footprints of the individual are also identical.
I have removed the assertions though as this has used [asizeof from the pympler library](`https://pympler.readthedocs.io/en/latest/library/asizeof.html)  - which would imply an addition to `text/requirements.txt`.

Reply:
> My reservations against `@classmethod` is that the user should not have to take care if a version is cached or not, but should just use always the same command. If this is not possible with calling `Dataset()`, we can maybe replace it by a function call, e.g. `dataset()` that then internally takes care of calling `Dataset()` or `Dataset.create()`. This might still look a little bit strange in the documentation and a user might wonder why not directly call `Dataset()`, but at least it's always the same command.

Without being API breaking, one could make the current `Dataset` private (i.e. rename into `_Dataset`) and have a simple wrapper class `Dataset` that uses the constructor `_Dataset.create`. 

Would that also be an option? One could do both.

Reply:
I have implemented it in b46e33d. I also is a little ugly in the wrapping class as I had to "copy" all `@staticmethods` in order to keep them being callable without having an instance.

I also am confident now that saving the entire objects using `@cached_properties` does not generate a massive memory footprint on hard disk.

I have made some experiments that could be shared.

Reply:
Cool, does it mean we are ready here yet?

Would be cool if you could add a description to this pull request and maybe also add the report on execution time there.

Would also be great to provide measurement times for the `@cached_properties` as you indicate that they are already faster even if they are calculated the very first time.
Looking in your benchmark table above `prop_extraction_tot` is by far the largest value, maybe we could still benefit by also pickle the results in the cache. But it's totally fine with me if we would do/investigate this in another pull request and focus here on getting the cache for `Dataset()` and use `@cached_properties` for the properties.

Reply:
I have run comparisons of cached and uncached object instance creation again (against 187 databases):
The implementation now always pre-calculates all properties.
The performance gain remains (the mean ratio suggests a factor of approx. 360). 

![image](https://github.com/audeering/audbcards/assets/14260675/04fe72f4-362f-4ac2-9e59-199004e09467)

I thought that redoing this analysis was useful in order to investigate the fix to the pickling problems that are discussed in the thread below.

Reply:
Great, can you add the information of the approx 360x performance gain to the description, then we can find it easier if we look for it at a later stage.